x/rollapp/keeper/rollapp.go

```diff
@@ -245,3 +245,109 @@ func (k Keeper) IsRollappStarted(ctx sdk.Context, rollappId string) bool {
 	_, found := k.GetLatestStateInfoIndex(ctx, rollappId)
 	return found
 }
+
+func (k Keeper) MarkRollappAsVulnerable(ctx sdk.Context, rollappId string) error {
+	return k.FreezeRollapp(ctx, rollappId)
+}
+
+// FreezeRollapp marks the rollapp as frozen and reverts all pending states.
+func (k Keeper) FreezeRollapp(ctx sdk.Context, rollappID string) error {
+	rollapp, found := k.GetRollapp(ctx, rollappID)
+	if !found {
+		return gerrc.ErrNotFound
+	}
+
+	rollapp.Frozen = true
+
+	k.RevertPendingStates(ctx, rollappID)
+
+	if rollapp.ChannelId != "" {
+		clientID, _, err := k.channelKeeper.GetChannelClientState(ctx, "transfer", rollapp.ChannelId)
+		if err != nil {
+			return fmt.Errorf("get channel client state: %w", err)
+		}
+
+		err = k.freezeClientState(ctx, clientID)
+		if err != nil {
+			return fmt.Errorf("freeze client state: %w", err)
+		}
+	}
+
+	k.SetRollapp(ctx, rollapp)
+	return nil
+}
+
+func (k Keeper) verifyClientID(ctx sdk.Context, rollappID, clientID string) error {
+	rollapp, found := k.GetRollapp(ctx, rollappID)
+	if !found {
+		return gerrc.ErrNotFound
+	}
+
+	var (
+		emptyRollappChannelID = rollapp.ChannelId == ""
+		emptyClientID         = clientID == ""
+	)
+
+	switch {
+	// both channelID and clientID are empty
+	case emptyRollappChannelID && emptyClientID:
+		return nil // everything is fine, expected scenario
+
+	// channelID is empty while clientID is not
+	case emptyRollappChannelID:
+		return fmt.Errorf("rollapp does not have a channel: rollapp '%s'", rollappID)
+
+	// clientID is empty while channelID is not
+	case emptyClientID:
+		return fmt.Errorf("empty clientID while the rollapp channelID is not empty")
```

Comment:
is this copy pasted?
seems weird to me like checking an invariant or something
when are these cases actually possible?